adds yammer search. closes #1454

plamber opened the pull request:
I am closing #1919 and create this new pull request

Implements #1454

waldekmastykarz commented on src/m365/yammer/commands/yammer-search.ts:
```diff
@@ -0,0 +1,214 @@
+
+  public getTelemetryProperties(args: CommandArgs): any {
+    const telemetryProps: any = super.getTelemetryProperties(args);
+    telemetryProps.search = typeof args.options.search !== 'undefined';
```
Since `search` is required, there is no point in tracking it in telemetry as it would always show up as set

waldekmastykarz commented on src/m365/yammer/commands/yammer-search.ts:
```diff
@@ -0,0 +1,214 @@
+
+  private getAllItems(logger: Logger, args: CommandArgs, page: number): Promise<void> {
+    return new Promise<void>((resolve: () => void, reject: (error: any) => void): void => {
+      let endpoint = `${this.resource}/v1`;
```
Since the endpoint is fixed, let's defined it as a `const` rather than composing of two pieces.

waldekmastykarz commented on src/m365/yammer/commands/yammer-search.ts:
```diff
@@ -0,0 +1,214 @@
+          const resultGroups = results.groups;
+          const resultUsers = results.users;
+
+          let continueprocessing = true;
```
Let's capitalize subsequent words for readability

```suggestion
          let continueProcessing = true;
```

waldekmastykarz commented on src/m365/yammer/commands/yammer-search.ts:
```diff
@@ -0,0 +1,214 @@
+            });
+        }
+        else {
+          logger.log(this.summary);
```
If I'm searching, I'd expect to see the results rather than stats. Is there a way to provide a simple overview of results matching the query, eg. title + URL, something that would be useful by itself?

plamber replied:
The challenge I see here is that we have different types of objects that are returned.

Messages, Topics, Groups, and Users.

The only common denominator I see is the ID and DisplayName. Everything else is somewhat custom. We could think of providing multiple tabular responses (do we have already something like that?) or to return following format:

ID, DisplayName, Type
where type is "GROUP, USER, TOPIC, MESSAGE"

Alternatively we could limit the search results to a type and return the whole object by passing a type parameter like "MESSAGE,TOPIC,USER,GROUP,SUMMARY" and then return the results. I would provide this option just for the "TEXT" output.

waldekmastykarz replied:
Is there a URL for each of these objects to open them in context of Yammer? What is the context of search? What would you want to do with the results next? I think answering these questions, will help us decide what kind of information we should display.

plamber replied:
There is just one endpoint that allows to perform a text search. It will then return a summary and all the objects found in that search. You do not have many other filtering capabilities. The added filtering capabilities and data output capabilities are then performed client side. 

I am still thinking that the JSON output is more valuable in terms of automation. I do not think that I would use the CLI to perform a basic text search. 

The only reasonable solution might be to provide some "enhanced" filtering capabilities with added parameters and filter with "MESSAGE,TOPIC,USER,GROUP,SUMMARY" the contents client side. If somebody wants to have everything in one shot, then JSON should be used.

waldekmastykarz replied:
I didn't mean the ability to filter, but rather having a link to the result, which you could open in the browser or include in a report. So when you execute the command, you get a table of title, type and URL. If you want all data for automation, you can grab JSON.

waldekmastykarz commented on src/m365/yammer/commands/yammer-search.ts:
```diff
@@ -0,0 +1,214 @@
+      },
+      {
+        option: '--limit [limit]',
+        description: 'Limits the results returned for each item category. Can only be used with the --output json option.'
```
If we adjust the output to present the results, then the limit option should apply in all output modes.

plamber replied:
See comment above. After we decided we can close this one

plamber replied:
Added the show command option to improve the handling

waldekmastykarz commented on src/m365/yammer/commands/yammer-search.ts:
```diff
@@ -0,0 +1,214 @@
+  }
+
+  public validate(args: CommandArgs): boolean | string {
+    if (args.options.search && typeof args.options.search !== 'string') {
```
What else can you pass as an argument?

plamber replied:
There is not much left

plamber replied:
We have now a show option to filter the type of resource in the text output

waldekmastykarz commented on src/m365/yammer/commands/yammer-search.ts:
```diff
@@ -0,0 +1,214 @@
+  search: string;
+  limit?: number;
+}
+
```
```suggestion

interface YammerSearchResponse {
  count: {
    groups: number;
    messages: number;
    topics: number;
    users: number;
  };
  groups: any[];
  messages: { messages: any[] };
  topics: any[];
  users: any[];
}

```

waldekmastykarz commented on src/m365/yammer/commands/yammer-search.ts:
```diff
@@ -0,0 +1,214 @@
+
+      request
+        .get(requestOptions)
+        .then((results: any): void => {
```
Since we're using properties of this object further in code, let's define its shape and not use `any` to avoid runtime errors.

```suggestion
        .then((results: YammerSearchResponse): void => {
```

waldekmastykarz commented on src/m365/yammer/commands/yammer-search.ts:
```diff
@@ -0,0 +1,214 @@
+      };
+
+      request
+        .get(requestOptions)
```
```suggestion
        .get<YammerSearchResponse>(requestOptions)
```

waldekmastykarz reviewed (changes requested):
Let's adjust a few things before we merge it in

plamber commented:
@waldekmastykarz I did some minor changes but we should still take a decision about the TEXT output

plamber commented:
Hi @waldekmastykarz,
added a new option "show" that allows you to specify which output to show when using the output text variant. The rest remains as defined.

I think the command is now much more usable than before. Thank you for your input.

br,
Patrick

waldekmastykarz commented:
> added a new option "show" that allows you to specify which output to show when using the output text variant. The rest remains as defined.

Why couldn't we allow users to search across all types of resources? Option like `show` could be useful but more as a filter that you can use to trim the results to the specific type, like a refiner. Default should be though results across all types, like you have in the UI.

plamber commented:
Do not agree with it. 

We get back different types with different columns in a complex structure. I do not like the idea to have mixed results in one tabular format. If I want tho slice and dice I can still use the json output

waldekmastykarz commented:
Compare it to the regular search experience in SharePoint: you get results of all content types mixed and matched in a single view by the lowest common denominator: title, description and URL. You can filter if you want, but that's an extra not a prerequisite.

plamber commented:
Hi ,
We only have id, title, url, and eventually a type as a result then. I would keep the other mode to have a filter by type.

Is this what you are referrring to?

waldekmastykarz commented:
I would indeed use the properties you mentioned in the default result set. I don't know if we should change output depending on the type of content selected in the filter. I think it could end up being confusing. Instead, I'd stick with consistent results across all types and then complete output with all properties in JSON mode.

plamber commented:
Hi @waldekmastykarz,
just added the consolidated output.

br,
Patrick

waldekmastykarz commented on docs/docs/cmd/yammer/yammer-search.md:
````diff
@@ -40,7 +40,7 @@ m365 yammer search [options]
+Returns the search result for the query `community`
 
 ```sh
 m365 yammer search --search "community"
````
If the `--show` option is required, as designated by `<>`, we should include it here, or running the command will fail

plamber replied:
Have fixed it in the documentation. It was changed to optional

waldekmastykarz commented on src/m365/yammer/commands/yammer-search.ts:
```diff
@@ -285,8 +290,48 @@ class YammerSearchCommand extends YammerCommand {
           } else { 
-            logger.log(this.summary);
+            let results: YammerBasicSearchResponse[] = [];
+            results = [...results, ...this.messages.map((msg) => {
```
Maybe it's just my limited understanding of the code, but how can you cast the same values to different types? It seems confusing

plamber replied:
the different sets of objects contain the same properties as strings. This output works fine

waldekmastykarz commented on docs/docs/cmd/yammer/yammer-search.md:
````diff
@@ -0,0 +1,77 @@
+Returns the groups found with the search query `community`
+
+```sh
+m365 yammer search --search "community" --show "groups"
````
We use lowercase `groups` here, but in the allowed values use `Groups` with capital `G`. Let's choose one format and use it consistently

plamber replied:
Changed everything to lowercase

waldekmastykarz reviewed (changes requested):
I'm looking at the default output and this is what I'm seeing:

<img width="916" alt="image" src="https://user-images.githubusercontent.com/11164679/101247919-8088e800-371c-11eb-86b7-09842f1c7b70.png">

A few things that we should do, is:
- decode the contents
- use the web URL so that users can click on the URL to see the particular result
- rename `--search` to `--queryText` (I saw you caught that we used `--query` in the spec which is a global option 👍 )

Other than that, I wonder if there is something that we could do, to make the output more human-readable. We have no control over the contents, and as they're so long, they break the table formatting. Since the ID is already included in the URL, we could drop the ID from the text output. Shall we also consider trimming the contents to let's say max 80 chars? Any other suggestions?

@pnp/cli-for-microsoft-365-maintainers any ideas?

plamber commented:
We have a similar issue with other yammer  commands. The output is in many situation difficult to manage.

Handling messages is anyway a challenge due the long text that might come back Trimming might help. I still do not like the output in non json format for messages. I find it useless

I can add trimming and I am fine in renaming the parameter. Any other suggestion before making a last change?

plamber commented:
@waldekmastykarz something that came into my mind. Do we have a way to specify a maxwidth for the text output columns and the contents change to a new line? This could help in such output scenarios

Any suggestions how to define a "dynamic output"?

I have a 49 inch widescreen. I might have less issues outputting multiple colums as somebody with smaller screens or resolution

waldekmastykarz commented:
For printing table output in text mode we use `easy-table`. I see it has the ability to define column width, but if I recall correctly, it doesn't support multi-line rows so the rendering will break. We could look for an alternative that does support multi-line rows or trim the content to a given length ourselves.

I never run terminal on full screen. In the past, we've been breaking lines in command's help on 80 chars. Now, we wrap lines following the terminal window width, but the table output remains challenging to get right in some cases where we have a lot of content to show.

@pnp/cli-for-microsoft-365-maintainers what's your thinking?

plamber commented:
@pnp/cli-for-microsoft-365-maintainers, do you have some input for us?

rabwill commented:
I vote for trimming and to keep table format

garrytrinder commented:
I agree with @rabwill ,that said, I rarely use the table output and pretty much always use JSON so if it’s not a perfect output, I wouldn’t be losing sleep over it.

plamber commented:
Hi @waldekmastykarz,
Performed following changes:
- web_url is shown instead of url in output
- renamed name to description
- trimmed output of messages to 80 chars 
- changed paramter to queryText instead of search

br

waldekmastykarz commented:
Hey @plamber, when testing the command, I noticed, that in the text output, when you use the `--show` option to show only specific type of results, you see different columns for each type of data:

<img width="912" alt="image" src="https://user-images.githubusercontent.com/11164679/102121430-44543680-3e44-11eb-8193-c41757c0881a.png">

<img width="910" alt="image" src="https://user-images.githubusercontent.com/11164679/102121506-5cc45100-3e44-11eb-8b61-ed3e3e49bdfa.png">

Haven't we decided to always show the same columns in the text mode and then show the full results in JSON output?

<img width="912" alt="image" src="https://user-images.githubusercontent.com/11164679/102121534-66e64f80-3e44-11eb-8e6c-c10a1b3e0b1d.png">

https://github.com/pnp/cli-microsoft365/pull/1944#issuecomment-735434758

plamber commented:
I most likely misunderstood this. Will change it

waldekmastykarz commented:
We're almost there. Appreciate your patience and sorry for the lack of clarity on my part ❤️

waldekmastykarz commented on src/m365/yammer/commands/yammer-search.spec.ts:
```diff
@@ -0,0 +1,622 @@
+    const options = ['summary', 'messages', 'users', 'topics', 'groups'];
+    options.forEach((option) => {
+      const actual = command.validate({ options: { queryText: '123', show: option } });
+      assert.strictEqual(actual, true);
```
When defining multiple asserts in a single test, always specify third parameter (message), that will help us understand which assert failed.

waldekmastykarz reviewed (approved):
Nicely done! I've adjusted a few things when merging the PR. 👏 

Tip for the next time, avoid using multiple asserts in one test without specifying the message, because it makes it hard for us to find out what's wrong exactly. 🙂

waldekmastykarz commented:
Merged manually. Thank you! 👏
